cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -320,25 +343,37 @@ private void processMetrics() {
 
         // Skip persisting if the records is empty
         if (records.isEmpty()) {
-          return;
+          return metricsProcessIntervalMillis; // Sleep for maximum time allowed if no new metrics are received
         }
 
         // Ensure there's only one thread can persist records and messageId's when the persistingFlag is false.
         // The thread set persistingFlag to true when it starts to persist.
         if (!persistingFlag.compareAndSet(false, true)) {
           LOG.trace("Cannot persist because persistingFlag is taken by another thread.");
-          return;
+          // Don't sleep if delay between current time and the last metric's time is larger than maxDelayMillis
+          if (System.currentTimeMillis() - lastMetricTimeMillis > maxDelayMillis) {
+            return 0;
+          }
+          // Sleep for metricsProcessIntervalMillis if the delay is no larger than maxDelayMillis
+          return metricsProcessIntervalMillis;
         }
+        Deque<MetricValues> recordsCopy = new LinkedList<>();
         try {
-          Deque<MetricValues> recordsCopy = new LinkedList<>();
           Map<TopicIdMetaKey, byte[]> topicMessageIdsCopy = new HashMap<>(topicMessageIds);
-          // TODO: (CDAP-8327) there is a risk of running out-of-memory if other threads keep writing to records
           Iterator<MetricValues> iterator = records.iterator();
-          while (iterator.hasNext()) {
+          while (iterator.hasNext() && recordsCopy.size() < queueSize) {
             recordsCopy.add(iterator.next());
             iterator.remove();
           }
-          persistRecordsMessageIds(recordsCopy, topicMessageIdsCopy);
+          long lastRecordTime = TimeUnit.MILLISECONDS.toSeconds(recordsCopy.peekLast().getTimestamp());
+          long delay = System.currentTimeMillis() - lastRecordTime;
```

Comment:
the delay should only depends on record seen by this thread. In fact, the logic should be quite simple as:

```java
long startTime = System.currentMillis();
// Fetch messages from TMS and put it to the shared queue
// During the looping over the fetched messages, also get the lastMetricTimeMillis
...
// If nothing was fetched, just return
if (records.isEmpty()) {
  return metricProcessIntervalMillis;
}

if (persistingFlag.compareAndSet(false, true)) {
  persistRecords(...);   // this is a new private method for the writer to persist data
}

long endTime = System.currentMillis();
if (endTime - lastMetricTimeMillis > maxDelayMillis) {
  // Don't sleep if falling behind
  return 0L;
} else {
  long timeSpent = endTime - startTime;
  return Math.max(0L, metricsProcessIntervalMillis - timeSpent);
}
```